File: Difficulty: Medium/Rod Cutting/rod-cutting.cpp
class Solution {
  public:
    int cutRod(vector<int> &price) {
        // code here
        int n = price.size();
        int length[n];
        for(int i=0;i<n;i++){
            length[i]=i+1;
        }
        int t[n+1][n+1];

        //intialization 
        for(int i=0;i<n+1;i++){
           for(int j=0; j<n+1; j++){
               if(i==0 || j==0){
                   t[i][j]=0;
               }
           }
        }

        //choice diagram
        for(int i=1; i<n+1; i++){
            for(int j=1; j<n+1; j++){
                if(length[i-1]<=j){
                    t[i][j] = max(price[i-1]+t[i][j-length[i-1]],t[i-1][j]);
                }
                else{
                    t[i][j]=t[i-1][j];
                }
            }
        }

        return t[n][n];
    }
};

File: Difficulty: Medium/Rod Cutting/rod-cutting.java
class Solution {
	public int solve(int[]price, int index, int rodLength, int[][]t) {
		if (rodLength == 0 || index == 0) {
			return 0;
		}
		
		if(t[index][rodLength] != -1){
		    return t[index][rodLength];
		} 
		// choice diagram
		int currentRodLength = index;
		int currentRodPrice = price[index - 1]; // 0-based indexing ke liye
		
		// Agar current piece rodLength se chota ya barabar hai, tabhi kaat sakte hain
		if (currentRodLength <= rodLength) {
			// Choice 1: Include (Kaato)
			// Profit mila + Rod ki length kam hui, LEKIN index same raha (Unbounded!)
			int include = currentRodPrice + solve(price, index, rodLength - currentRodLength,t);
			
			// Choice 2: Exclude (Mat Kaato)
			// Profit 0 + Rod length same rahi, lekin next choti length try karenge (index - 1)
			int exclude = solve(price, index - 1, rodLength,t);
			
			t[index][rodLength] = Math.max(include, exclude);
		}
		
		else {
			// MAJBOORI: Piece size rod length se bada hai -> Sirf Exclude kar sakte hain
			t[index][rodLength] = solve(price, index - 1, rodLength,t);
		}
		
		return t[index][rodLength];
	}
	public int cutRod(int[] price) {
		// code here
		int totalRodLength = price.length;
		int totalOption = price.length;
		
		int [][] t = new int[totalRodLength+1][totalRodLength+1];
		for(int[] row : t){
		    Arrays.fill(row,-1);
		}
		
		
		return solve(price, totalOption, totalRodLength,t);
	}
}
